# Credit Risk Evaluator

In [1]:
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

## Retrieve the Data

The data is located in the Challenge Files Folder:

* `lending_data.csv`

Import the data using Pandas. Display the resulting dataframe to confirm the import was successful.

In [2]:
# Import the data
train_df = pd.read_csv(Path('Resources/lending_data.csv'))
train_df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

In [3]:
test_df = pd.read_csv(Path('Resources/lending_data.csv'))
test_df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

Prediction:
Logicistic Regression is typically used to financial setting to analyse credit scores,so my prediction is Logistic Regression will perform better because it is used to predict financial, biological or sociological data in supervised machine learning. Random Forest however is ensemble-based learning algorithm, comprised of collections of de-correlated decision trees.



## Split the Data into Training and Testing Sets

In [4]:
# Split the data into X_train, X_test, y_train, y_test
y_train = train_df["loan_status"]
X_train = train_df.drop(columns = ["loan_status"])
X_train.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

In [5]:
y_test = test_df["loan_status"]
X_test = test_df.drop(columns = ["loan_status"])
X_test.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

In [6]:
X_train = pd.get_dummies(X_train)
X_test = pd.get_dummies(X_test)
print(f"Train: {X_train.shape}, Test: {X_test.shape}")

Train: (77536, 7), Test: (77536, 7)


In [7]:
for col in X_train.columns:
    if col not in X_test.columns:
        X_test[col]=0
print(f"Train: {X_train.shape}, Test: {X_test.shape}")

Train: (77536, 7), Test: (77536, 7)


## Create, Fit and Compare Models

Create a Logistic Regression model, fit it to the data, and print the model's score. Do the same for a Random Forest Classifier. You may choose any starting hyperparameters you like. 

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the designated markdown cell.

In [8]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression
clf_logistic  = LogisticRegression(solver='lbfgs').fit(X_train, y_train) 

print(f"Training Data Score: {clf_logistic.score(X_train, y_train)}")
print(f"Testing Data Score: {clf_logistic.score(X_test, y_test)}");

Training Data Score: 0.9920037144036319
Testing Data Score: 0.9920037144036319


In [9]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier
clf_forest = RandomForestClassifier(random_state=42, n_estimators=200).fit(X_train, y_train)

print(f'Training Score: {clf_forest.score(X_train, y_train)}')
print(f'Testing Score: {clf_forest.score(X_test, y_test)}')

Training Score: 0.9969304581097813
Testing Score: 0.9969304581097813


In [13]:
# Scale the data
scaler = StandardScaler()
scaler.fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [14]:
# Train the Logistic Regression model on the scaled data and print the model score
logisticRegr_scaled = LogisticRegression(
    solver='lbfgs', 
    max_iter=100,
    random_state=0
)
logisticRegr_scaled.fit(X_train_scaled, y_train)
print("LogisticRegression scaled score: ", logisticRegr_scaled.score(X_test_scaled, y_test))

LogisticRegression scaled score:  0.9941317581510524


In [15]:
# Train a Random Forest Classifier model on the scaled data and print the model score
randomForestClass_scaled = RandomForestClassifier(random_state=0)
randomForestClass_scaled.fit(X_train_scaled, y_train)
print("RandomForestClassifier scaled score: ", randomForestClass_scaled.score(X_test_scaled, y_test))

RandomForestClassifier scaled score:  0.9969304581097813


# Results
LogisticRegression unscaled score:  0.9920037144036319
LogisticRegression scaled score:  0.9941317581510524
RandomForestClassifier unscaled score:  0.9969304581097813
RandomForestClassifier scaled score:  0.9969304581097813 

    
#Conclusion
I think RandomForestClassifier model is better for this data set because it remains same in both scaled and unscaled data
while there is slight change in values in Logistic Regression model.